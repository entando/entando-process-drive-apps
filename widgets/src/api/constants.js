export const DOMAINS = {
  PDA: process.env.REACT_APP_DOMAIN || '',
  APP_BUILDER: process.env.REACT_APP_APP_BUILDER_DOMAIN || '',
};

export const IS_MOCKED_API = process.env.REACT_APP_MOCKED_API === 'true';

export const MOCK_API_DELAY = 800;

export const LOCAL = process.env.REACT_APP_LOCAL === 'true';

export const METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE',
};

// Inject ur access_token if you are trying to reach the APP-BUILDER API locally
export const TOKENS = {
  APP_BUILDER: '0694821a5bdc2229aac820e8a773669a',
};
