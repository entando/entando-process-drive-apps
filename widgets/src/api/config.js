import { DOMAIN, IS_MOCKED_API, MOCK_API_DELAY } from 'api/constants';
import utils from 'utils';
import config from 'mocks/config';

const get = async widgetCode => {
  if (IS_MOCKED_API) {
    await utils.timeout(MOCK_API_DELAY);
    return config;
  }

  const url = `${DOMAIN}/configs/${widgetCode}`;
  const response = await fetch(url);

  return response.json();
};

export default {
  get,
};
