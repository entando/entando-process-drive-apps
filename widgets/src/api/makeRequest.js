import { IS_MOCKED_API, MOCK_API_DELAY, DOMAINS, METHODS } from 'api/constants';
import utils from 'utils';

const getParams = queryParams =>
  Object.keys(queryParams)
    .filter(k => queryParams[k] !== undefined && queryParams[k] !== '')
    .map(k => `${encodeURIComponent(k)}=${encodeURIComponent(queryParams[k])}`)
    .join('&');

export default async ({
  domain,
  uri,
  method,
  mockResponse = {},
  useAuthentication,
  body,
  headers = {},
  queryParams,
  forceMock,
}) => {
  if (IS_MOCKED_API || forceMock) {
    await utils.timeout(MOCK_API_DELAY);
    return mockResponse;
  }

  const url = `${domain || DOMAINS.PDA}${uri}${queryParams ? `?${getParams(queryParams)}` : ''}`;

  const requestHeaders = useAuthentication
    ? new Headers({
        'Content-Type': 'application/json',
        Authorization: `Bearer ${localStorage.getItem('token')}`,
        ...headers,
      })
    : new Headers(headers);

  const configs = {
    method: method || METHODS.GET,
    ...(body ? { body } : {}),
    headers: requestHeaders,
  };

  const response = await fetch(url, configs);
  if (!response.ok) {
    const payload = await response.json();
    throw new Error(
      payload.message
        ? payload.message
        : `Bad Response from server: ${response.status} ${response.statusText}`
    );
  }

  const responseHeaders = response.headers.get('Content-Type');

  if (responseHeaders && responseHeaders.includes('xml')) {
    return response.text();
  }
  if (responseHeaders && responseHeaders.includes('octet')) {
    return response;
  }
  return response.json();
};
