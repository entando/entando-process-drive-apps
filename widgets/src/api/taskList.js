import { IS_MOCKED_API, MOCK_API_DELAY, SERVICE } from 'api/constants';
import getTasks from 'mocks/taskList/api.mock';
import utils from 'utils';

const get = async (connection, page, pageSize, sortedColumn, sortOrder, filter) => {
  if (IS_MOCKED_API) {
    await utils.timeout(MOCK_API_DELAY);
    return getTasks(page, pageSize, sortedColumn, sortOrder, filter);
  }

  const url = `${SERVICE.URL}/connections/${connection}/tasks?sort=taskId`;
  const token = localStorage.getItem('token');
  const response = await fetch(url, {
    method: 'get',
    headers: new Headers({
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    }),
  });

  return response.json();
};

export default {
  get,
};
