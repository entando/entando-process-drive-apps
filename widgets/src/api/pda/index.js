import makeRequest from 'api/makeRequest';
import { SERVICE, METHODS } from 'api/constants';

import MOCK_CONNECTIONS from 'mocks/pda/connections';
import MOCK_PROCESSES from 'mocks/pda/processes';

export const getConnections = async () =>
  makeRequest({
    domain: SERVICE.URL,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: MOCK_CONNECTIONS,
    useAuthentication: false,
  });

export const getProcess = async connection =>
  makeRequest({
    domain: SERVICE.URL,
    uri: `/connections/${connection}/processes/definitions`,
    method: METHODS.GET,
    mockResponse: MOCK_PROCESSES,
    useAuthentication: false,
  });
