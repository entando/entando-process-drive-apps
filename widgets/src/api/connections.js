import { METHODS, DOMAINS } from 'api/constants';

import { CONNECTIONS } from 'mocks/taskList/configs';
import makeRequest from 'api/makeRequest';

// eslint-disable-next-line import/prefer-default-export
export const getConnections = async () =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: CONNECTIONS,
    useAuthentication: false,
  });
