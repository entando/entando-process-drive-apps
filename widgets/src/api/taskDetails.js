import { SERVICE, METHODS } from 'api/constants';
import mockedTask from 'mocks/taskDetails/task';

import { CONNECTIONS, PROCESS, GROUPS, COLUMNS } from 'mocks/taskList/configs';
import makeRequest from 'api/makeRequest';

export const getTask = async (connection, taskId) => {
  return makeRequest({
    domain: SERVICE.URL,
    uri: `/connections/${connection}/tasks/${taskId}`,
    method: METHODS.GET,
    mockResponse: mockedTask,
    useAuthentication: true,
  });
};

// Configs

export const getConnections = async () =>
  makeRequest({
    domain: SERVICE.URL,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: CONNECTIONS,
    useAuthentication: false,
  });

export const getProcess = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/processes/definitions`,
    method: METHODS.GET,
    mockResponse: PROCESS,
    useAuthentication: true,
  });

export const getGroups = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/groups`,
    method: METHODS.GET,
    mockResponse: GROUPS,
    useAuthentication: true,
  });

export const getColumns = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/tasks/columns`,
    method: METHODS.GET,
    mockResponse: COLUMNS,
    useAuthentication: true,
  });
