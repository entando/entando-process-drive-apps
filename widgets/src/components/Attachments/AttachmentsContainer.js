/* eslint-disable no-console */
import { MuiThemeProvider as ThemeProvider } from '@material-ui/core/styles';
import React from 'react';
import PropTypes from 'prop-types';
import List from '@material-ui/core/List';
import withStyles from '@material-ui/core/styles/withStyles';
import Button from '@material-ui/core/Button';

import { DOMAINS, LOCAL } from 'api/constants';
import { getAttachments, saveAttachment, deleteAttachment } from 'api/pda/attachments';
import { Typography } from '@material-ui/core';
import { getPageWidget } from 'api/app-builder/pages';
import SimpleDialog from 'components/common/SimpleDialog';
import Notification from 'components/common/Notification';
import theme from 'theme';
import AttachmentsSkeleton from './AttachmentsSkeleton';
import AddAttachmentModal from './AddAttachmentModal';
import Attachment from './Attachment';

const styles = {
  footer: {
    textAlign: 'right',
  },
  empty: {
    border: 'solid 1px #eee',
    marginTop: 8,
    marginBottom: 2,
    background: 'white',
    borderRadius: 3,
    padding: 18,
    textAlign: 'center',
    height: 52,
  },
};

class AttachmentsContainer extends React.Component {
  state = {
    attachments: [],
    loading: true,
    dialogOpen: false,
    connection: '',
    notification: {},
  };

  componentDidMount = async () => {
    const { serviceUrl } = this.props;

    if (!LOCAL) {
      // set the PDA domain to the URL passed via props
      DOMAINS.PDA = serviceUrl;
    }

    const config = await this.fetchConfigs();
    this.setState({ connection: config.knowledgeSource }, async () => {
      const attachments = await this.fetchAttachments();
      this.setState({ attachments, loading: false });
    });
  };

  fetchConfigs = async () => {
    const { pageCode, frameId } = this.props;

    try {
      const widgetConfigs = await getPageWidget(pageCode, frameId, 'ATTACHMENTS');
      if (widgetConfigs.errors.length) {
        throw widgetConfigs.errors[0];
      }
      if (!widgetConfigs.payload) {
        throw new Error('No configuration found for this widget');
      }

      const { config } = widgetConfigs.payload;
      return config;
    } catch (error) {
      this.handleError(error);
    }
    return {};
  };

  fetchAttachments = async () => {
    const { connection } = this.state;
    const { taskId } = this.props;

    let payload = [];

    try {
      const attachments = await getAttachments(connection, taskId);
      if (attachments.errors.length) {
        throw attachments.errors[0];
      }

      payload = attachments.payload;
    } catch (error) {
      if (!error.message.includes('404')) {
        this.handleError(error);
      }
    }
    return payload;
  };

  toggleDialog = () => {
    this.setState(state => ({ dialogOpen: !state.dialogOpen }));
  };

  handleUpload = async files => {
    const { connection } = this.state;
    const { taskId } = this.props;
    console.log('begin upload of files:');

    this.setState({ loading: true });
    try {
      const promises = files.map(async file => {
        await saveAttachment(connection, taskId, file);
      });

      const responses = await Promise.all(promises);
      if (responses.length) {
        const attachments = await this.fetchAttachments();
        this.setState({ attachments });
      }
    } catch (error) {
      this.handleError(error);
    }

    this.setState({ loading: false }, this.toggleDialog);
  };

  handleDelete = item => async () => {
    const { connection } = this.state;
    const { taskId } = this.props;
    try {
      this.setState({ loading: true });
      await deleteAttachment(connection, taskId, item.id);
      const attachments = await this.fetchAttachments();
      this.setState({ attachments });
    } catch (error) {
      this.handleError(error);
    } finally {
      this.setState({ loading: false });
    }
  };

  handleError = error => {
    this.setState({
      notification: { message: typeof error === 'object' ? error.message : error, type: 'error' },
    });
  };

  handleCloseNotifications = () => {
    this.setState({ notification: {} });
  };

  render() {
    const { attachments, loading, dialogOpen, connection, notification } = this.state;
    const { classes, taskId } = this.props;

    return (
      <ThemeProvider theme={theme}>
        <div>
          <Typography variant="h3">Attachments</Typography>
          {loading ? (
            <AttachmentsSkeleton rows={3} />
          ) : attachments.length ? (
            <List>
              {attachments.map(item => (
                <Attachment
                  key={item.id}
                  item={item}
                  onDelete={this.handleDelete}
                  downloadLink={`${DOMAINS.PDA}/connections/${connection}/tasks/${taskId}/attachments`}
                />
              ))}
            </List>
          ) : (
            <div className={classes.empty}>
              <Typography>There is no attachment available for this task.</Typography>
            </div>
          )}
          <div className={classes.footer}>
            <Button variant="outlined" color="primary" onClick={this.toggleDialog}>
              Add
            </Button>
          </div>
        </div>
        <SimpleDialog
          title="Add new Attachment"
          open={dialogOpen}
          body={<AddAttachmentModal onUpload={this.handleUpload} onClose={this.toggleDialog} />}
          onClose={this.toggleDialog}
          maxWidth="md"
        />
        <Notification
          message={notification.message}
          type={notification.type}
          onClose={this.handleCloseNotifications}
        />
      </ThemeProvider>
    );
  }
}

AttachmentsContainer.propTypes = {
  classes: PropTypes.shape({
    footer: PropTypes.string,
    empty: PropTypes.string,
  }),
  taskId: PropTypes.string.isRequired,
  serviceUrl: PropTypes.string,
  pageCode: PropTypes.string,
  frameId: PropTypes.string,
};

AttachmentsContainer.defaultProps = {
  classes: {},
  serviceUrl: '',
  pageCode: '',
  frameId: '',
};

export default withStyles(styles)(AttachmentsContainer);
