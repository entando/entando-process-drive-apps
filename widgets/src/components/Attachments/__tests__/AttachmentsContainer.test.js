import React from 'react';
import { render, wait } from '@testing-library/react';

import { DOMAINS } from 'api/constants';
import WIDGET_CONFIGS from 'mocks/app-builder/pages';
import PARAMS from 'mocks/app-builder/widgets';
import MOCKED_ATTACHMENTS_RESPONSE from 'mocks/pda/attachments';
import AttachmentsContainer from '../AttachmentsContainer';
import 'mocks/i18nMock';

describe('<AttachmentsContainer />', () => {
  it('renders snapshot correctly', async () => {
    const configUrl = `${DOMAINS.APP_BUILDER}/api/pages//widgets/`;
    const connection = 'kieStaging';
    const { taskId } = PARAMS.ATTACHMENTS;
    const attachmentsList = `/connections/${connection}/tasks/${taskId}/attachments`;

    fetch
      .once(JSON.stringify(WIDGET_CONFIGS.ATTACHMENTS))
      .once(JSON.stringify(MOCKED_ATTACHMENTS_RESPONSE));

    const { container } = render(<AttachmentsContainer taskId={taskId} />);

    await wait(() => expect(container).toMatchSnapshot());

    expect(fetch.mock.calls.length).toBe(2);
    expect(fetch.mock.calls[0][0]).toEqual(configUrl);
    expect(fetch.mock.calls[1][0]).toEqual(attachmentsList);
  });
});
