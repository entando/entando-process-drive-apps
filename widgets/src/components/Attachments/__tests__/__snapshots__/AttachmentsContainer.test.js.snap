// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AttachmentsContainer /> renders snapshot correctly 1`] = `
<div>
  <div>
    <h3
      class="MuiTypography-root MuiTypography-h3"
    >
      Attachments
    </h3>
    <ul
      class="MuiList-root MuiList-padding"
    >
      <li
        class="MuiListItem-container"
      >
        <div
          class="MuiListItem-root Attachment-listItem-127 MuiListItem-gutters MuiListItem-secondaryAction"
        >
          <div
            class="MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm-1 4l6 6v10c0 1.1-.9 2-2 2H7.99C6.89 23 6 22.1 6 21l.01-14c0-1.1.89-2 1.99-2h7zm-1 7h5.5L14 6.5V12z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root Attachment-truncate-128"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
            >
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                download=""
                href="/connections/kieStaging/tasks/28@mortgage-process_1.0.0-SNAPSHOT/attachments/2/download"
              >
                55954396_2301387639927701_986123889866178560_o.jpg
              </a>
            </span>
          </div>
        </div>
        <div
          class="MuiListItemSecondaryAction-root"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </li>
      <li
        class="MuiListItem-container"
      >
        <div
          class="MuiListItem-root Attachment-listItem-127 MuiListItem-gutters MuiListItem-secondaryAction"
        >
          <div
            class="MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm-1 4l6 6v10c0 1.1-.9 2-2 2H7.99C6.89 23 6 22.1 6 21l.01-14c0-1.1.89-2 1.99-2h7zm-1 7h5.5L14 6.5V12z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root Attachment-truncate-128"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
            >
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                download=""
                href="/connections/kieStaging/tasks/28@mortgage-process_1.0.0-SNAPSHOT/attachments/4/download"
              >
                55954396_2301387639927701_986123889866178560_o.jpg
              </a>
            </span>
          </div>
        </div>
        <div
          class="MuiListItemSecondaryAction-root"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </li>
    </ul>
    <div
      class="AttachmentsContainer-footer-1"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  
</div>
`;
